### ISSUE TITLE
Line is still connected when eliminating output

### ISSUE THREAD

**Issue by juliandavidmr** [retejs/rete#234]
When an output is added and then multiple connections are made, then I eliminate that output but the connections still exist

![error mapper](https://user-images.githubusercontent.com/4544102/50164117-d81fb600-02af-11e9-93ce-8ac57269927a.gif)

**Comment by Ni55aN** [comment]
Show me your code :fox_face:

**Comment by juliandavidmr** [comment]
I have been following the error, it seems that the plugin "connection-plugin" does not work in certain cases. Failure occurs when "output" does not exist:

https://github.com/retejs/connection-plugin/blob/master/src/picker.js#L30

![error connection](https://user-images.githubusercontent.com/4544102/50167122-6dbe4400-02b6-11e9-9182-2f4b86ab17b7.png)

**Comment by juliandavidmr** [comment]
It eliminates the "Output" of a component and the "_output" of the plugin continues initialized, we must find a way to reset the value of "_output" (from connection-plugin) when an Output is removed from a component.

**Comment by Ni55aN** [comment]
I think that you forgot to remove connections before removing of Output (`editor.removeConnection(connection)`)

**Comment by juliandavidmr** [comment]
I already did it, it works but there are residues of the connection, for example the sockets created with the directive, until now I think it may be the error.

**Comment by juliandavidmr** [comment]
I have tried in all ways, the connections are removed but the socket never work again.

```ts
    const originNode = this.getOriginNodes()[attribute.originAt];
    const output = originNode.outputs.get(Array.from(originNode.outputs.keys())[attribute.indexAttribute]);

    if (output.hasConnection()) {
      output.connections.map(c => {
        const attr = c.input.node.inputs.get(c.input.key);
        this.editor.view.removeConnection(c);

        attr.connections.map(c2 => {
          if (c2.output.key === output.key) {
            c2.remove();
          }
        });

        this.editor.removeConnection(c);
        c.remove();
      });

      /*const line = document.querySelector(
        `.connection, [class^=input-${ c.input.name.split(' ')[0] }], [class^=output-${ c.output.name.split(' ')[0] }]`
      );

      if (line) {
        line.remove();
      }*/
      // c.remove();
    }
```

**Comment by Ni55aN** [comment]
`this.editor.view.removeConnection(c);` and `c2.remove();` interferes to remove the connections correctly. Therefore `this.editor.removeConnection(c);` can no longer get connections to remove them, since they have already been removed from the list, but connection-render and other plugins have not know it.

`editor.removeConnection`  will be enough
```js
    input.connections.slice().map(editor.removeConnection.bind(editor));
    node.removeInput(input);
```

http://jsfiddle.net/Ni55aN/vmxdcLbq/91/

**Comment by juliandavidmr** [comment]
I have managed to reproduce the problem:

http://jsfiddle.net/anlijudavid/pxgms4jz/22/

![error rete](https://user-images.githubusercontent.com/4544102/50185369-91e54980-02e5-11e9-83cd-cfd4da301d1f.gif)

**Comment by Ni55aN** [comment]
You are not calling `editor.removeConnection` to remove connection

**Comment by Ni55aN** [comment]
All operations to remove objects must be performed via `editor`

**Comment by borja010** [comment]
Good evening,

I had the same error before and I tried a lot of ways without being able to fix it. 

I was able to resolve the issue by cloning the node and re-doing the connections.

**Comment by juliandavidmr** [comment]
Hello @borja010, I have also tried a lot of ways and I have not been able to solve it, I also thought of a solution like yours but I thought it was an extreme case, but after not knowing what happens I will follow your advice, thank you very much. 



_Sent from my Moto G(4) using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by Ni55aN** [comment]
@juliandavidmr you need to update the node and connections after `node.removeOutput(output);`
```js
  // node.update(); //for vue render
  node._alight.scan()
  editor.view.updateConnections({node})
```

**Comment by Ni55aN** [comment]
@borja010 this should work
```js
   input.connections.slice().map(editor.removeConnection.bind(editor));
   node.removeInput(input);

        
    node.update();
    editor.view.updateConnections({node})
```

**Comment by juliandavidmr** [comment]
I have applied the connections and node update changes, but the editor completely fails, see the gif below:

![86c5cfcc-596d-4e5a-804a-b30ef90b52d2](https://user-images.githubusercontent.com/4544102/50228772-c0f7cb80-0376-11e9-8ff6-65eeaf937b21.gif)

**Comment by Ni55aN** [comment]
I can not make a pause in the gif :)

**Comment by borja010** [comment]
When I was debugging the same issue.  I identify that the issue only happens with the output that is below from the one that is deleted and when you click the output, the picker detects the information from the output that was deleted.

**Comment by juliandavidmr** [comment]
That's right, the problem is that the socket of the eliminated Output continues to exist, on the other hand the connections plugin continues with the problem that I describe in XXXX

> It eliminates the "Output" of a component and the "_output" of the plugin continues initialized, we must find a way to reset the value of "_output" (from connection-plugin) when an Output is removed from a component.

**Comment by Ni55aN** [comment]
I have changed render plugin and reproduced issue http://jsfiddle.net/Ni55aN/vmxdcLbq/102/

**Comment by juliandavidmr** [comment]
I just saw the example, it seems that this is the same problem

**Comment by Ni55aN** [comment]
This issue was fixed in reder-plugin v0.2.6. The problem was that it recreated the HTMLElement in the loop with a sockets, so the old one was lost

```js
input.connections.slice().map(editor.removeConnection.bind(editor));
node.removeInput(input);
     

await node.update(); // await is a little hack that allows to execute the next command on next frame (insteadof Vue.nextTick)
editor.view.updateConnections({node})
```

http://jsfiddle.net/Ni55aN/vmxdcLbq/104/

**Comment by juliandavidmr** [comment]
Now it worked, thank you very much @Ni55aN . You have saved my life.
